phoenix-core/src/main/java/org/apache/phoenix/iterate/DefaultParallelScanGrouper.java

```diff
@@ -17,46 +17,79 @@
  */
 package org.apache.phoenix.iterate;
 
+import java.sql.SQLException;
 import java.util.List;
 
+import com.google.common.base.Preconditions;
+import org.apache.hadoop.hbase.HRegionLocation;
 import org.apache.hadoop.hbase.client.Scan;
 import org.apache.phoenix.compile.QueryPlan;
+import org.apache.phoenix.compile.StatementContext;
 import org.apache.phoenix.schema.PTable;
 import org.apache.phoenix.schema.PTable.IndexType;
 import org.apache.phoenix.schema.SaltingUtil;
+import org.apache.phoenix.schema.TableRef;
 import org.apache.phoenix.util.ScanUtil;
 
 /**
  * Default implementation that creates a scan group if a plan is row key ordered (which requires a merge sort),
- * or if a scan crosses a region boundary and the table is salted or a local index.   
+ * or if a scan crosses a region boundary and the table is salted or a local index.
  */
 public class DefaultParallelScanGrouper implements ParallelScanGrouper {
-	
-	private static final DefaultParallelScanGrouper INSTANCE = new DefaultParallelScanGrouper();
 
-    public static DefaultParallelScanGrouper getInstance() {
-        return INSTANCE;
-    }
-    
-    private DefaultParallelScanGrouper() {}
-
-	@Override
-	public boolean shouldStartNewScan(QueryPlan plan, List<Scan> scans, byte[] startKey, boolean crossedRegionBoundary) {
-		PTable table = plan.getTableRef().getTable();
-		boolean startNewScanGroup = false;
-        if (!plan.isRowKeyOrdered()) {
-            startNewScanGroup = true;
-        } else if (crossedRegionBoundary) {
-            if (table.getIndexType() == IndexType.LOCAL) {
-                startNewScanGroup = true;
-            } else if (table.getBucketNum() != null) {
-                startNewScanGroup = scans.isEmpty() ||
-                        ScanUtil.crossesPrefixBoundary(startKey,
-                                ScanUtil.getPrefix(scans.get(scans.size()-1).getStartRow(), SaltingUtil.NUM_SALTING_BYTES), 
-                                SaltingUtil.NUM_SALTING_BYTES);
-            }
-        }
-        return startNewScanGroup;
+  private static DefaultParallelScanGrouper INSTANCE = new DefaultParallelScanGrouper();
```

Comment:
I don't think that DefaultParallelScanGrouper can be a singleton with the state of context and tableName inside of it.